test/Chainweb/Test/Cut.hs

```diff
@@ -88,31 +91,93 @@ import qualified Test.QuickCheck.Monadic as T
 import Chainweb.BlockCreationTime
 import Chainweb.BlockHash
 import Chainweb.BlockHeader
-import Chainweb.BlockHeight
 import Chainweb.ChainId
+import Chainweb.ChainValue
 import Chainweb.Cut
-import Chainweb.Difficulty (checkTarget)
+import Chainweb.Cut.Create
 import Chainweb.Graph
 import Chainweb.Test.Utils (genEnum)
-import Chainweb.Time (Micros(..), Time, TimeSpan, second)
+import Chainweb.Time (Micros(..), Time, TimeSpan)
+import qualified Chainweb.Time as Time (second)
 import Chainweb.Utils
 import Chainweb.Version
+import Chainweb.Version.Utils
 import Chainweb.WebBlockHeaderDB
 
 import Data.CAS.RocksDB
 
 import Numeric.Additive
 import Numeric.AffineSpace
 
+-- -------------------------------------------------------------------------- --
+-- Utils
+
+type TestHeaderMap = HM.HashMap BlockHash BlockHeader
+
+testLookup
+    :: MonadThrow m
+    => TestHeaderMap
+    -> ChainValue BlockHash
+    -> m BlockHeader
+testLookup db (ChainValue _ h) = fromMaybeM MissingParentHeader $ HM.lookup h db
+
+type GenBlockTime = Cut -> ChainId -> Time Micros
+
+-- | Block time generation that offsets from previous chain block in cut.
+--
+offsetBlockTime :: TimeSpan Micros -> GenBlockTime
+offsetBlockTime offset cut cid = add offset
+    $ maximum
+    $ fmap (_bct . _blockCreationTime)
+    $ HM.insert cid (cut ^?! ixg cid)
+    $ cutAdjs cut cid
+
+arbitraryBlockTimeOffset
+    :: TimeSpan Micros
+    -> TimeSpan Micros
+    -> T.Gen GenBlockTime
+arbitraryBlockTimeOffset lower upper = do
+    t <- genEnum (lower, upper)
+    return $ offsetBlockTime t
+
+-- | Solve Work. Doesn't check that the nonce and the time are valid.
+--
+solveWork :: HasCallStack => WorkHeader -> Nonce -> Time Micros -> SolvedWork
+solveWork w n t =
+    case runGet decodeBlockHeaderWithoutHash $ BS.fromShort $ _workHeaderBytes w of
+        Nothing -> error "Chainwb.Test.Cut.solveWork: Invalid work header bytes"
+        Just hdr -> hdr
+            & blockNonce .~ n
+            & blockCreationTime .~ BlockCreationTime t
+
+            -- After injecting the nonce and the creation time will have to do a
+            -- serialization roundtrip to update the Merkle hash.
+            --
+            -- A "real" miner would inject the nonce and time without first
+            -- decoding the header and would hand over the header in serialized
+            -- form.
+            --
+            & encodeBlockHeaderWithoutHash
+            & runPut
+            & runGet decodeBlockHeaderWithoutHash
+            & fromJuste
+            & SolvedWork
```

Comment:
`&` is defined in `base`. 

I find this sequence easier to read in top down direction. Using `$` would result in the exact same code lines just in reversed order (which makes inlining comments less readable because the reader has to change direction and it's not that obvious if the comment refers to the line above or below.